discover-providers.js

```diff
@@ -58,15 +58,21 @@ function createJsonFileDiscoverProvider(hostsFile) {
             return;
         }
 
-        var hosts;
-        try {
-            hosts = JSON.parse(fs.readFileSync(hostsFile).toString());
-        } catch (e) {
-            callback(e);
-            return;
-        }
+        fs.readFile(hostsFile, function onFileRead(err, data) {
+            if (err) {
+                callback(err);
+                return;
+            }
 
-        return callback(null, hosts);
+            try {
+                var hosts =JSON.parse(data.toString());
+                callback(null, hosts);
```

Comment:
if the callback throws an error it is caught by the try catch resulting in an extra call of the `callback`.

Can you move the invocation outside the `try`?